### ISSUE TITLE
 SAP Cloud SDK - Questions about upcoming SDK V3 Release

### ISSUE THREAD

**Issue by drvup** [SAP/cloud-sdk-js#3051]
Hey guys, 

Within the last 18 months we created more than 40 different VDM packages and more than 15 nodeJs APPs within the SAP Cloud SDK on a customer project. We do have some kind of data-lake implemented with CAP application and surrounding ingestion- and extraction-layer applications within nodeJs apps and the usage of SAP Cloud SDK.

On our agenda we have still the left-over topic about updating (older, stable) applications, using SDK v1, to the newest version of v2. Before we're starting with that, we wanna make sure about the release planing of v3, because maybe it does make sense to wait until the release of the newest major version is going to be published ;-)?

Beside that, we wanna get more details about the release roadmaps of the SDK and your release cycles as well. 
Making it more transparent does help us, our customers and the community to get better capabilities to plan our release cycles accordingly to the ones of our key dependency: your beautiful SDK.

Thanks for a short feedback and your support.

Kind Regards,
Cedric

**Comment by jjtang1985** [comment]
Hi @drvup ,

Thank you very much for your adoption and support as a user!

> Before we're starting with that, we wanna make sure about the release planing of v3

Please check our release policy [here](https://sap.github.io/cloud-sdk/docs/js/release-policy).

> because maybe it does make sense to wait until the release of the newest major version is going to be published ;-)?

It depends on your use cases. 
For example, for v2 upgrade, if you only use the `http-client` instead of `vdm`, there is not much effort.

In general, by following our current [upgrade guide](https://sap.github.io/cloud-sdk/docs/js/guides/upgrade-to-version-2), you should be able to upgrade from v1 to v2 with maximal 1 day (our goal).
Sure, there will be a new one for v3 upgrade, which might also take 1 day.
However, doing v2 upgrade and v3 upgrade together might be risky, just like a CI/CD pipeline:
- if your pipeline is always green, perfect
- if you commit frequently, and fix issues timely, every step should be relatively easy
- if you commit huge amounts of files or have a yearly release schedule without testing properly before, you might spend more time for fixing the issue
 
Therefore, I recommend doing it step by step, because eventually you have to cover both upgrade guide.

> we wanna get more details about the release roadmaps of the SDK and your release cycles as well.

- The release cycles can be found in the link above.
- The scope of v3 needs to be finalised and will be published later. I can share the current status (might be changed):
  - Allow user to define a middleware and use it for the http requests (also odata/openapi). A typical use case would be the resilience middleware, where e.g., `retry`/`circuit breaker`/`timeout` can be enabled. See ADR [here](https://github.com/SAP/cloud-sdk-js/blob/main/knowledge-base/adr/0031-resilience-options.md)
  - Improve our OData and OpenAPI generator for better performance and UX

Let me know if it helps.

Thank you very much.
BR,
Junjie

**Comment by drvup** [comment]
Hey @jjtang1985 ,

thanks for your fast response. Always appreciated :) 

Thanks for pointing me to the right path on the documentation about release schedules / cycles etc. - I do not know why I did not see it on the page before :/

Bigger thanks to point me to the knowledge-base of this repo! Really nice things there. Anyhow, great way to share and keep some knowledges visible to everyone!

=> I was also thinking the same way like you, talking about upgrading from v1 to v3 via v2 first. I made the experience that it's not done within 1 day updating applications to use and adopt sdk v2. The reason for it is the other architecture of de- / serializers and the schema (I used to love it after some time and I see the effort within custom de- / serializers in use). 

From customer POV, it's important to highlight to keep the upgrading effort as low as possible (they often still compare it with ABAP world -.-).

Thanks for sharing some insights about the upcoming changes, looking forward to get more details :)

Thanks and Kind regards,
Cedric

**Comment by FrankEssenberger** [comment]
@Junjie thanks for the great answer. I would like to add a few comments:
- We waited rather long for the first major version v1 -> v2. This lead to a rather big number of breaking changes because a lot of legacy has accumulated. We plan to make releases more [frequent](https://sap.github.io/cloud-sdk/docs/js/release-policy). This means more often less pain, which is in general a better feeling then very rarely a big pain.
- We tried to avoid the word `migration` or `update` because in the onPrem world this often means a big project with hunderts of working days for customers. Here it is really just a npm library update.
- If possible use TypeScript with the SDK because it will show errors directly at compile time and with a high test coverage you create confidence that your update is not breaking something.
- Since version 6 CAP also uses the SDK 2.0 so I would recommend to align the dependencies to have small node_modules.

Best
Frank
